chapter{Network Models}\label{network_models}


In this chapter we present and formally describe two classes of network models that were developed during this study.
\textbf{Species-collectors networks} (SCNs) are built based on associations between collectors and the species they've recorded during their careers, whereas \textbf{collectors coworking networks} (CWNs) describe direct collaborative associations between collectors when recording specimens in field.

Although structurally distinct, models here presented provide complementary perspectives on the recording behavior of collectors from a given species occurrence dataset. 
From SCNs we retrieve information on which collectors have recorded which species and, conversely, which species were recorded by which collectors. 
CWNs, on the other hand, allow us to investigate which collectors team up with whom during fieldwork, although here species are not represented as entities.

As both models were elaborated based on a social networks analytics framework, we start by reviewing some general concepts from the network science and graph theory domains.



% ======================
% ======================
% Network Science Theory
% ----------------------

\section{Network Science: A theoretical background}

Network Science refers to a relatively new domain of scientific investigation, aiming to describe emergent properties and patterns from complex systems of interacting entities.
Such relational systems are naturally represented as networks, where interactions are represented as pairwise connections (\textit{links}) between entities (\textit{nodes}) and assume particular semantics depending on the nature of the modeled phenomenon.
The rise of this field is strongly associated to recent advances in information technology, which provided scientists with novel tools for collecting, storing and processing data from many knowledge domains more efficiently and in larger scales.
Although a variety of networked systems in many disciplines had been studied long before that, technological advances allowed us to model real-world systems in much more details, from large volumes data that are often public or easily accessible for the investigator.

In a classic $1999$ paper which has inspired many researchers to engage into the field of network science, \textit{Albert-László Barabási}, \textit{Hawoong Jeong} and \textit{Réka Albert} built a model of the World-Wide Web from data collected by a web crawler \cite{Albert1999}. 
Their model represented the web as a set of interconnected documents, where a connection between document $a$ and document $b$ existed if either $a$ contained hyperlinks pointing to $b$ (outgoing links of $a$) or if $a$ was referred by $b$ through hyperlinks (incoming links of $a$).
By exploring some of the model topological features, they estimated that although the web is composed by a huge number of documents (by that time there were around $8 \times 10^8$ documents online), two randomly chosen documents are, in average, separated by as few as 19 links. 

Although surprising, this finding was in fact consistent with a generative model proposed by \textit{Watts} and \textit{Strogatz} in $1998$ \cite{Watts1998}.
In that paper authors argued that real-world networked systems are neither completely randomly nor completely regularly connected, and demonstrated that more realistic topologies could be derived by combining features from both extremes.
Following this approach they were able to generate network models in which nodes were separated from others by very few connections, even for very large systems.
This phenomenon is known as the \textit{small-world}, and became popular after the experimental work of \textit{Stanley Milgram} \cite{Milgram1969} as the ``six degrees of separation''.
The concept referred to the finding that two arbitrary individuals in the world are separated by at most 6 connections, following an acquaintances chain.
 
Besides small-world networks, other generative models have been proposed in literature to explain mechanisms by which real world networks grow and acquire their particular topologies.
A \textit{preferential attachment mechanism} was proposed by \textit{Barabási} as ruling network growth by preferentially connecting new nodes to those in the network which already have many connections \cite{Albert2002}. 
This phenomenon is also referred to as ``the rich gets richer'' effect, and allows the appearance of few heavily connected nodes in the network, called\textit{hubs}. 
The majority of nodes in the network are, in contrast, very poorly connected, and thus unlikely to be linked to be included in new edges.
Networks with such characteristics are known as \textit{scale-free} networks.

In random network models, introduced by \textit{Paul Erdos} and \textit{Alfred Rény} \cite{erdos1959random}, nodes are expected to connect to each other by chance, independently of the number of connections one already has.
According to \textit{Erdos} and \textit{Rény} $G(N,L)$ model, random networks are generated by creating $N$ distinct nodes and connecting them randomly and in a pairwise fashion by a total of $L$ links. 
If real-world networks behaved like random networks, a topology where few hubs coexisted with a massive amount of very poorly connected nodes would be extremely unlikely to observe. 
Therefore links are not established randomly in real-world networks. 
A plenty of system-dependent factors are assumed to play important roles in links creation, making non-trivial topological features to emerge. 
We refer to such networks as textit{complex networks}, which are the core of network science.


%% Examples of networks in biology, computer science...
Network science has been applied to many knowledge domains, including computer science, biophysics, neuroscience, and biodiversity.
Here we focus on applications in the field of biodiversity.

% Present some network models already developed applied to the biodiversity domain
%% Ecological interactions networks

%%% Co-occurrence networks
%%% Metagenomics
%%% Climatic changes
%%% Challenge: Reliable data is hard to be obtained

%% The herbarium exchangers network
%%% A prosopological approach, where social interactions between collectors is modeled


% Network models represented as graphs
Network models are suitable for structuring relational data, where nodes are expected to interact with each other through connections (or links). 
Networked systems as relational models, where entities interconnect by links.
Networks are mathematically represented as \textit{graphs}.
 
 
Next we briefly review some fundamental definitions from graph theory.
For the purpose of this work this is not intended to be a thorough review on the field, but rather a quick overview for readers that are not familiarized with the field. 
For a slightly more comprehensive introduction we recommend the reader to refer to \textit{Barabási}'s and \textit{Newman}'s books on network science, \cite{Barabasibook,Newman2010b}.

% ============
% Graph Theory
% ------------
\subsection{General concepts on Graph theory}

Intuitively, \textbf{graphs} are mathematical structures composed by discrete objects holding pairwise connections to each other. 
In graph theory terminology, objects are referred to as \textit{vertices} and the connections between them as \textit{edges}.
Graph structure provides a natural representation for networked systems for a couple of reasons.
First, graphs allow representing entities and their interactions in a structured way, as a ``map of connections'' through which information flows.
Second, graph theory provides a bunch of well established data structures, metrics and algorithms for systematically representing, characterizing and exploring graph topologies.
Finally, the availability of a considerable amount of graph visualization tools and techniques helps analysts to obtain insights from network structure, by allowing them to focus on different aspects of the system depending on their interest.

% Graph formal definition, undirected/weighted graphs
Graphs are formally defined as a pair $G=(S,E)$, where $S$ is the set of vertices that compose the graph; and $E$ is the graph's edges set. 
%
Edges necessarily involve pairs of vertices, and are thus represented as $(u,v)$, for $u,v \in S$.
Pairs of vertices that are directly connected by an edge are said to be \textit{adjacent}, and the set of all vertices that are adjacent to a given vertex composes its \textit{neighborhood}.
%
Depending on the nature of the modeled connections, edges can be either classified as \textit{undirected} or \textit{directed}.
Undirected edges model symmetric connections, in which case information is allowed to flow equally in both directions.
There are situations, however, in which information necessarily flows from a source to a target vertex, thus making connections asymmetric.
Those are represented in the graph as directed edges.
Graphs composed by directed edges are known as \textit{digraphs} (or \textit{directed graphs}), whilst \textit{undirected graphs} only have undirected edges.
In undirected graphs we should note that edges $(u,v)$ and $(v,u)$ are equivalent.
%
Additionally, connections in a networked system can be modeled as either being all identical or distinct in terms of their strengthness or relevance, which is incorporated in the graph as edges \textit{weight}.
Graphs are also classified as either weighted on unweighted, based on the nature of their edges.
%
Given the nature of the networks modeled in the context of this project, we will specifically focus on properties of undirected weighted graphs for the rest of this section.


% Graph representations
Graphs can be computationally represented using a variety of data structures, the most appropriate one heavily depending on the set of algorithms one expects to run against the graph.
Here we review some of the most common of them.
%
\textit{Edges lists} are perhaps the simplest and most intuitive graph representation, and consists of a list storing all edges as ordered pairs $(u,v)$, for $u,v \in S$.
Given its simplicity, this representation is more appropriate for situations in which the user directly interacts with the data, such as when manually creating or editing graphs without the aid of specialized software. 
This representation is also compact enough to be adopted for graph storage and data exchange in some situations.
%
Most algorithms, however, perform better by using alternative designs such as \textit{adjacency lists} or \textit{adjacency matrices}. 
%
\textbf{Adjacency matrices} are a way to represent graphs in matrix form, in which elements store adjacencies between pairs of vertices.
An adjacency matrix $A$ is defined as having non-zero entries $A_{i,j}$\textit{ iff } $(i,j)\in E$ which, in weighted graphs, correspond to the weight assigned to edge $(i,j)$.
Given the equivalence between edges $(i,j)$ and $(j,i)$ in undirected graphs, adjacency matrices in this case are always symmetric, meaning information is stored redundantly in its structure.
%
Besides, the size of an adjacency matrix grows quadratically with the number of vertices in the graph, which would make loading and processing large graphs in computer memory problematic.
However, real-world networks have been observed to be naturally sparse, meaning that only a very small percentage of all possible edges do in fact exist.
Adjacency matrices could thus be optimized for storage and performance by adopting standard sparse matrix representations such as compressed row storage (CRS) \ref{Saad2003}.
An alternative way to represent sparse networks is to use the \textit{adjacency list} representation, which in short consists of storing a list of neighbors for every single vertex in the network. 
% TODO:<include figure: graph where nodes have attributes and edges are weighted; simple graph with its adj. matrix>


% Definitions and metrics
As previously mentioned, one of the main benefits of adopting graph models for representing networks is the availability of a whole set of well-established metrics and definitions that allow analysts to characterize the topology of their systems. 
Below we review some of the most basic ones.

\paragraph*{Path}
A path can be thought of as a chain of adjacent vertices composing a route through which information flows in a graph.
For undirected graphs we say there exists a \textit{path} of length $n$ between a pair of vertices if they are mutually reachable by following a sequence of $n$ edges.
Pairs of vertices are said to be \textit{connected} if at least one path exists between them.
A pair of vertices can be connected through more than one path, each having distinct lengths.
The \textit{distance} between two nodes is the length of the shortest path connecting them.
We can calculate the average path length for a network as the average path length between all pairs of connected nodes in the network.
% Diameter of the network

\paragraph*{Connected component}
A connected component is a subset of vertices in the graph in which all included vertices are \textit{connected} to every other one by at least one path.
Moreover, this subset is required to be maximal, meaning that all vertices in the network for which the inclusion property holds should also be included in the component. 
A network graph can be composed by many distinct connected components which are completely disconnected from each other, the largest of which is called the \textbf{giant component}.

\paragraph*{Clique}
A clique is a structure composed by a subset of vertices in which all included vertices are \textit{adjacent} to each other.
The set of edges that compose a clique are obtained by computing the pairwise combination of all its $n$ vertices, being the total number of edges equivalent to the binomial coefficient $\binom{n}{2}$.


\paragraph*{Attributes}
Attributes allow representing non-topological features that are relevant in the context of the modeled system, and can be optionally assigned to both edges and vertices in a graph.
For example, we might want to model a social network in which individuals name, age and gender information are relevant for our analyses.
Those features, which assume distinct values for each individual in the network, is typically stored as vertices attributes.
Similarly, edges attributes store particularities regarding each individual connection
In thesis, any numerical attribute could be used for weighting the edges.

\paragraph*{Degree}
The degree of a vertex in a undirected graph is given by the total number of vertices that are adjacent to it or, in other words, the length of its neighborhood.
We compute the degree of a particular vertex $i$ as 
$
k_i = \sum_j^{n} A_{i,j}
$,
where $A_{i,j}$ are the elements in the graph's adjacency matrix containing edges weights $(i,j)$.
The \textit{average degree} ($\langle k \rangle$) is a global property of the network, computed by averaging the degree values for each individual vertex $i$.
$
\langle k \rangle = \frac{1}{N} \sum_{i=1}^N k_i
$

\paragraph*{Degree distribution}
The probability distribution of vertices degrees over a graph is known as its \textit{degree distribution}.
From degree distribution we retrieve the probability $p_k$ that a randomly selected vertex from the graph has degree equal to $k$.
In random graphs degree distribution is typically well approximated by a poisson model, which peaks at $\langle k \rangle$.
Thus vertices with degree equal $\langle k \rangle$ are most likely to occur, whereas those with very high and very low degrees are very unlikely. 
In most real-world networks, however, a majority of low-degree nodes coexist with few hubs.
Such a scenario better described by a\textit{power law}, where $p(k) \sim k^{-\alpha}$.



%% Bipartite graphs
%%%% The bipartite constraint: formalization: Nodes sets are DISJOINT (no intersection) ; and INDEPENDENT (no adjacent vertices within any set); 
%%%% Biadjacency matrix
%%%% Bipartite projection
%%%%%% bipartite tradeoff : summarization vs information loss
%%%%%% bipartite projections edges set is usually very dense -> The importance of adding weights to edges in bipartite projections
%%%%%% weaker connections are then filtered out
\paragraph*{Bipartite graphs}
Bipartite graphs, also known as bigraphs or two-mode graphs, are a special class of graphs composed by two distinct sets of vertices $U$ and $V$, with the constraint that no vertices within the same set are allowed to be adjacent to each other. 
We define bipartite graphs as triples 
$
B = (U, V, E) \mbox{, }
$
where $E$ is the set of edges between vertices from $U$ and $V$.
In sets theory terms, $U$ and $V$ are both \textit{disjoint} and \textit{independent} sets.
This means vertices must be assigned to exactly one vertices set and, moreover, all edges in $E$ necessarily connect vertices from opposite sets. 
Such features make bipartite graphs particularly useful for representing interactions that only make sense to exist between entities of different classes. 

For bipartite graphs the adjacency matrix is known as \textit{biadjacency matrix}
In case the graph is unimodal the adjacency matrix is a squared symmetrical matrix.
For bipartite graphs, however, this is not necessarily true. It depends on the size of each vertices set. We refer to it as the \textbf{biadjacency matrix}.

As most graph algorithms and metrics in literature are primarily designed for one-mode graphs, bipartite graphs ...
One way to summarize bipartite 
% Bipartite projection

  \begin{figure}[h!]
  	\centering
    \includegraphics[width=0.5\linewidth]{figures/bipartite_general.png}
    \caption{(a) General aspect of a bipartite graph. All vertices in the graph belong to exactly one of $U$ and $V$ vertices sets. In addition edges are only established between vertices from distinct sets. (b) Bipartite projections. Projections onto each node set are constructed by linking together vertices that are at a length-2 distance in the bipartite graph, while omitting vertices from the other set.}
    \label{fig:bipartite_general}
  \end{figure}



\subsection{The Social Networks framework}
% Mark Granovetter paper (1973)
In the context of this work we refer to \textit{social networks} as systems in which people interact with each other through some type of social tie. 
Under the framework of network science, those systems are modeled as networks in which people are represented as nodes and social ties as links. 
In fact, a variety of social network models within many distinct application domains have been proposed in literature. 
We describe some of them for making the reader a more solid intuition before we present our models developed throughout our work.


\textit{Affiliation networks} are a particular type of social networks in which actors are associated to events, groups or institutions by being members or participating in them \cite{Borgatti2015}. Social ties between individuals are usually derived from such models by considering that individuals belonging to the same groups or attending the same events are more likely to become acquainted than those who are not co-affiliated to any of those. 
A classic example of this kind of network is the movie actors network, which was used as an empirical example in \textit{Watts} and \textit{Strogratz}'s work \cite{Watts1998} while describing the small-world property in real-world networks.
This network was built from the Internet Movie Database (IMDB \footnote{\url{http://www.imdb.com/}}), linking actors to movies they have starred in.

% exemplify models

Although the semantics of relationships modeled in each of these networks (with different analytic purposes), the structures used to represent them are shared. A common feature shared by those models is their bipartite structure, allowing  

The bipartite structure of SCNs make them structurally similar to affiliation networks, with some examples in the social networks literature.

One classic affiliation network model is the films and actors network.

Lambiotte models listeners and music

Another analogous example are the scientific papers coauthorship networks in which authors are linked to papers they have authored \cite{Newman2004}



Whereas SCNs represent ..., other structures might arise when analyzing which collectors record together. 
The model presented in the next section covers this.



% ===========================
% ===========================
% Species-collectors Networks
% ---------------------------

\section{Species-collectors Networks}

In this section we describe Species-collectors Networks (SCNs) as relational models for structuring associations between collectors and species.
We first give an overview on the semantics of the relationships we propose to model and how they can be derived from a species occurrence dataset.
Next we define attributes and operations for SCNs that facilitate obtaining domain-specific insights from it.
Finally we compare our proposed model to other analogous though different systems in literature.
% Illustrate the concept; 

\subsection{General Description}
%% Model semantics
Species-collectors networks describe relationships of type ``\textbf{collector} samples \textbf{species}'' or, conversely, ``\textbf{species} is sampled by \textbf{collector}'' (Figure \ref{fig:scn_general}). Such relationships, which necessarily involve both a collector and a species, are structured as links in the network.
The network is thus composed by collectors holding links to every single species they have ever recorded or alternatively, species holding links to every collector who have ever recorded them.
An important semantic aspect of this model worth emphasizing is that here we model collectors recording species rather than specimens. As exposed elsewhere in this text, the term \textit{species} refers to aggregations of taxonomically similar \textit{specimens}, which are the actual individuals that are collected. Thus, while collectors are represented at the individual level (each collector is a person), species are instead entities composed of groups of individuals, and must be uniquely included in the network as such.
This means no particular collector or species should be represented more than once in the model, even though they might have been observed in multiple records in dataset.

  \begin{figure}[h!]
  	\centering
    \includegraphics[width=.7\linewidth]{figures/network_models/scn_generalaspect.pdf}
    \caption{Multiple perspectives of a Species-collectors Network model (SCN).
    (a) Unprojected network, where collectors (green nodes) are linked to the species (red nodes) they've recorded. The total number of records of a given species by some collector is reflected in the strength of their link. (b) SCN projection onto the species set. Species are linked together if they've been collected by common collectors. Link strength is proportional to the number of collectors two species share. (c) SCN projection onto the collectors set. Collectors are linked together if they've recorded species in common. Link strength is proportional to the number of species two collectors share. 
    Link strength for both projections were obtained using the \textit{simple weighting} rule (eq. \ref{eq:simple_weighting}), and are graphically displayed as edges thickness. Nodes sizes reflect collectors' and species' degrees in each perspective.}
    \label{fig:scn_general}
  \end{figure}
  
%% Model specification
As networks are formally described as graph structures, we represent both collectors and species entities as nodes, though belonging to distinct classes.
Links are represented as edges.
Moreover, as relationships here modeled can only possibly exist between collectors and species we also impose a constraint that all edges in the network must necessarily connect nodes from distinct classes. In other words, nodes can be divided into two disjoint sets, with the condition that no edges in the same set are adjacent.
This best matches the description of a bipartite network
$$ SCN = (S_{col},S_{sp},E) \mbox{ ,}$$
where $S_{col} = \{u_1, u_2, ..., u_n \}$ is the nodes set representing the collectors group; $S_{sp}=\{v_1,v_2, ..., v_m\}$ is the nodes set representing the species group; and $E$ is the set of undirected edges between members of $S_{col}$ and $S_{sp}$.

The graph can also be represented as a rectangular biadjacency matrix $A^{n\times m}$ for which $a_{ij}\neq 0$ \textit{iff} $(u_i,v_j) \in E$. Values of non-zero $a_{ij}$ elements are set to the number of times edges $(u_i,v_j)$ occur in the network, as described below. 
For large and sparse SCNs, storage and operations on the graph object can be optimized by using sparse matrix designs for $A$.


  
%% Network construction
\subsection{Model construction from data}
We use basically two fields to build a SCN model from a species occurrence dataset. First, we need a collectors field, containing the names of all collectors that were responsible for the record; and the species field, storing the species identity assigned to the specimen in the record.
Following Darwin Core terms standard\footnote{\url{http://rs.tdwg.org/dwc/terms/}}, we should expect to find collectors names in the \textit{recordedBy} field; and the species name in a field named \textit{species}. As not every biological collection dataset uses Darwin Core standards though, these fields might be occasionally found under different names.

The network is built up from the dataset in an iterative process.
For each row in the dataset each individual collector included in the record establishes an additional link to the collected species.
As at least one additional link is necessarily either created or reinforced for each new row, the construction process guarantees that no disconnected species or collector nodes can possibly exist in the network model.
In this process nodes and edges are either created in case they did not previously exist in the graph; or have their occurrence counts increased in case they did.


%% < add an illustration >

We keep the record of the number of times each link occurs in the network by setting a \textit{count} attribute to them, which is initially set to $1$ and is increased by one every time a new occurrence of the link is observed. Link strength is proportional to this attribute, and is graphically represented by edges thickness in Figure \ref{fig:scn_general}.
The more often a particular collector records a particular species, the stronger gets the link between them.
Edges' \textit{count} values are stored in the biadjacency matrix $A$, and thus the value of element $a_{ij}$ is the number of times the edge $(u_i, v_j)$ occurs. 
The adjacency matrix for the example SCN network in Figure \ref{fig:scn_general} is thus
$$
A =
\kbordermatrix{
& sp1 & sp2 & sp3 & sp4 & sp5 & sp6 & sp7 & sp8 & sp9 & sp10 \\
col1 & 1 & 1 & 1 & 2 & 1 & 0 & 0 & 0 & 0 & 0 \\
col2 & 1 & 1 & 1 & 1 & 0 & 1 & 0 & 0 & 0 & 0 \\
col3 & 1 & 2 & 1 & 0 & 1 & 1 & 0 & 0 & 3 & 0 \\
col4 & 0 & 0 & 0 & 1 & 0 & 0 & 4 & 1 & 0 & 0 \\
col5 & 0 & 0 & 0 & 3 & 0 & 0 & 0 & 0 & 0 & 1 \\
}.
$$
An homonymous attribute is also set to graph nodes, which is increased whenever a new link involving the node is either added or strengthened. As a result the node's \textit{count} attribute keeps a record of how many times a given species or collector occurs in the dataset.
The reader should note, however, that \textit{count} attributes for nodes and edges are conceptually distinct and are not to be confused.


% SCN Definitions

\subsection{Definitions}
Given an overall description on the structure and semantics of the SCN model, we now define a set of attributes and operations that provide higher-level abstractions for dealing with the system here modeled. By using such field-domain abstractions we potentialize the data exploration process, eventually making it more insightful for the analyst.
We introduce both the \textit{species bag} and the \textit{quorum vector} as specific attributes of collectors and species.
Moreover, we define the process of taxonomy aggregation as a model summarization routine for grouping together species nodes into higher-rank taxa.

\paragraph{Species bag.} 
The entire set and counts of species a collector has recorded in a dataset, which can be thought as a collector's species signature, composes his/her \textit{species bag}. This attribute is therefore exclusively derivable for collectors nodes.
As species bags are directly obtained as row-vectors of the graph's biadjacency matrix, they are a convenient structure for comparing collectors in terms of the composition of their records.
For that task a high variety of well-known distance algorithms for vectors in literature can be readily applied.
The species bag $\sigma$ for collector $u_i$ is defined as

$$
\sigma_{u_i} =  \begin{bmatrix}
a_{i 1}, a_{i 2}, ..., a_{i m}
\end{bmatrix}  \quad ,
$$
where $m$ is the length of the species set and each $a_{i j}$ is the total number of records of species $v_j$ by collector $u_i$. The sum of all elements in a collector's species bag, which is equivalent to the vector's \textit{l1 norm} $||\sigma_{u_i}||_1$, corresponds to the total number of records for that collector.

 
\paragraph{Quorum.} 
The entire set and counts of collectors who have recorded a particular species in a dataset comprise its \textit{quorum}, an exclusive attribute of species nodes. 
This concept can be thought as the inverse of a species bag, being the collectors signature of a species. 
The quorum vector $\iota$ of a species $v_j$ is directly obtained from the graph's biadjacency matrix as the $j^{th}$ column-vector 

$$
\iota_{v_j} = \begin{bmatrix}
a_{1 j}, a_{2 j}, ..., a_{n j}
\end{bmatrix} \quad ,
$$
where $n$ is the length of the collectors set and each $a_{i j}$ is the total number of times collector $u_i$ has recorded species $v_j$. 
The total number of occurrences of species $v_j$ in the entire dataset can be obtained as the sum of all elements in its quorum vector $ || \iota_{v_j} ||_1$.


\paragraph{Taxonomic aggregation and resolution.}
In some contexts it might be desired to simplify SCNs by grouping species nodes into higher taxonomic ranks (or levels), such as \textit{genus} or \textit{family}. This process is defined as \textit{taxonomic aggregation}, and is performed by 
($i$) obtaining a grouping of species using some taxonomic rank; 
($ii$) obtaining quorum vectors for each species; 
($iii$) summing up quorum vectors for all species in each group;
($iv$) building a new SCN model, aggregated on rank T. 
The SCN's \textit{taxonomic resolution} is the taxonomic rank at which species are aggregated in the model. For the sake of model interpretability, all nodes in $S_{sp}$ must necessarily be represented as taxons belonging to the same rank as the SCN's taxonomic resolution.

For a more formal description let $G_T = \{g_1,g_2,..., g_n\}$ denote a taxonomic grouping at rank $T$, containing a set of $n$ rank-$T$ taxa.  In addition, let each taxon $g_i \in G_T$ itself be a set of nodes $S_{sp}^{(i)} \subseteq S_{sp}$, with the conditions that there are no empty $S_{sp}^{(i)}$ and that every node $v \in S_{sp}$ is a member of exactly one set  from $ \{ S_{sp}^{(1)}, S_{sp}^{(2)}, ..., S_{sp}^{(n)} \}$.
Such a grouping rule makes $G_T$ a partition of $S_{sp}$, and thus the entire set $S_{sp}$ can be recreated by simply computing the union of elements in $G_T$. This guarantees that no entities are duplicated or eliminated on aggregations using it.

We then use grouping $G_T$ for obtaining quorum vectors for each of its taxa $g_i \in G_T$, which will be represented as nodes in the new aggregated graph. Quorum vectors are computed as $ \iota_{g_i} := \sum_j \iota_{v_j}  $ for $v_j \in S_{sp}^{(i)}$. Finally, the rank-$T$ aggregated  graph $SCN_T=(G_T,S_{col},E)$ is created from a biadjacency matrix, which is constructed by stacking quorum vectors for each taxon $g_i$ as row-vectors. The set of collectors nodes remain the same in the aggregated graph.

% The PICI model (Lambiotte2005)
%% Collective effects acting on individuals with similar interests
%% Individual mechanisms, pushing collectors towards their particular interests, establishing their collecting niche

% Temporal edges


% SCN projections
\subsection{Projections}

Bipartite projections on each one of the SCN's nodes sets allows one to investigate indirect associations two entities from the same class might have with each another, as intermediated by a third entity from the opposite class. 
Figure \ref{fig:scn_general} illustrates projections of a SCN onto the species set (b) and the collectors set (c).
In overall, each projection gives us complementary perspectives of transitive relationships in the SCN, either from the collectors or species point of view.

From a species-centric perspective (Figure \ref{fig:scn_general}b), connections are formed between species having been recorded by at least one collector in common, with link strength proportional to the number of different collectors they share. 
Although collectors are used during projection for determining the existence of links between species they are not represented as nodes in this projection.
In general strongly connected species can be interpreted as being both included in the species bags of many collectors, whereas weakly connected or isolated species are seldom or never recorded by the same collectors.
% Modules in species projections reveal species that are more intensely associated with each other than with others from outside the module
The second perspective (Figure \ref{fig:scn_general}c) is collectors-centric, in that only collectors are represented as nodes whilst species are omitted. 
Analogously to the species-centric perspective, here collectors are linked together if they have recorded at least on species in common, with link strength depending on the number of shared species between them. 
From this perspective we could identify collectors having similar recording profiles.

As previously discussed in this text, projections are a mechanism for summarizing bipartite into more convenient one-mode graphs, where only one class of entity is represented. % TODO: Remember to add this in the projection section
Projections, however, come with the cost of information loss, as any relationships or attributes of nodes from the omitted set are not represented in the projection \cite{Borgatti1997}. % TODO: Add reference
Moreover, relevant associations between entities eventually become obfuscated by others of lower relevance, as projections tend to generate graphs that are much denser than the original bipartite model \cite{Lambiotte2005}.
Choosing an appropriate weighting rule for the aspects one wants to investigate is thus detrimental for separating relevant from less-relevant associations, so that the latter ones can be subsequently removed by applying weighting filters.
Below I first describe the simplest weighting rule with its limitations and, in sequence, some alternatives rules for overcoming them.
% https://doi.org/10.1016/j.physa.2006.12.021 <- The effect of weight on community structure of networks

\paragraph*{Simple weighting.}
This rule assigns weights to links between pairs of collectors ($u_s$ and $u_t$) or species ($v_s$ and $v_t$) by simply counting the total number of species collectors share on their species bags or the total number of collectors species share in their quorum vector. The rule is mathematically expressed as:
\begin{equation} \label{eq:simple_weighting}
\begin{split}
w_{(u_s, u_t)} &= \sum_{j=1}^{m} \delta(\sigma^{(j)}_{u_s}, \sigma^{(j)}_{u_t})\mbox{ , for the projection onto }S_{col}\mbox{ ;}\\
w_{(v_s, v_t)} &= \sum_{i=1}^{n} \delta(\iota^{(i)}_{v_s}, \iota^{(i)}_{v_t})
\mbox{ , for the projection onto }S_{sp}\mbox{, where}
\end{split}
\end{equation}
$n = |S_{col}|$, $m = |S_{sp}|$, $\sigma^{(i)}$ and $\iota^{(i)}$ are the $i^{th}$ element of a species bag and a quorum vector, respectively; and $\delta(u,v)=1$ if both $u$ and $v$ are non-zero and $0$ otherwise.

In order to obtain the weights for every pairs of projected nodes more efficiently we can use a vectorized implementation of this rule. First we derive a $n\times m$ logic matrix $A_{bool}$ from the SCN biadjacency matrix $A$ by simply replacing its non-zero elements by ones. 
Then a $n \times n$ adjacency matrix with edges weights for the $S_{col}$ projection is obtained by calculating the dot product $A_{bool} A_{bool}^T$.
Conversely, for the $S_{sp}$ projection, the $m \times m$ weights matrix is obtained by calculating $A_{bool}^T A_{bool}$.

The simple weighting rule has an important limitation when applied to SCNs.
It arises from the fact that the weight assigned to edges linking pairs of nodes in the projection only reflects the number of distinct intermediate neighbors from the complementary set they shared in the non-projected graph. The number of times each species is recorded by each collector is therefore ignored while computing the strength of links in the projections, underestimating the importance of recurrent relationships.
Consequently this weighting rule tends to make very prolific and generalist collectors or very attractive species strongly connected to many others in a disproportional way, as an effect of their high degrees in the non-projected model. The opposite happens in the case of specialized nodes, which typically hold fewer --- although recurrent --- distinct links to their neighbors. 
In order to reduce these effects two alternative rules are proposed below.

\paragraph*{Additive weighting.}
This rule is a slight modification of the simple weighting rule, in that it also considers the total number of times entities interact through each neighbor-intermediated path in the non-projected network. The rule is expressed using the same equations from (\ref{eq:simple_weighting}), but changing the $\delta$ function to
 
$$\delta(u,v) = 
\begin{cases}
\frac{u+v}{2} &  \mbox{if both } u \mbox{ and } v \mbox{ are non-zero ,}\\
0 & \mbox{otherwise}
\end{cases}
$$
In case every distinct path in the non-projected SCN only occur once then both simple and additive weighting rules lead to the same result. % TODO: Test this assumption 
This modified rule potentializes the effect of recurring edges from the non-projected graph on computing edges weights in the projection, thus reducing weighting asymetries from generalist and specialized nodes.

However, this approach still has a drawback in that nodes which have high degrees in the non-projected graph tend to become much more strongly connected with themselves in the projection than average-degree ones, simply by the fact that they have many more connections than average.
Additionally, without a superior limit for the $\delta$ function it turns out to be hard to determine a proper threshold when filtering relevant from non-relevant edges. The next weighting rule is designed to reduce the effects of nodes degrees on their edges' weights, outputting values which are bounded to the $[0,1]$ interval.

\paragraph*{Species bag / Quorum similarity.}
% This is called "Structural Equivalence": Nodes have ties to common third-parties {Borgatti2015}
% One characteristic is that in practice collectors holding similarity values very close to one tend to be those with very few records in the dataset. More experienced collectors can have higher similarities with some collectors, but it is usually not very high.
This weighting rule uses a similarity (or correlation) matrix that is computed for each projection of the SCN. Edges' weights are given by the similarity between their nodes. The similarity matrix for the collectors and species projections are constructed by computing the  \textit{cosine similarity} of species bags and quorum vectors  for each pair of nodes in the respective projection. 
The \textit{species bag similarity} for collectors $u_s$ and $u_t$; and the \textit{quorum vector similarity} for species $v_s$ and $v_t$ are defined as

\begin{equation}
\begin{split}
sim(\sigma_{u_s},\sigma_{u_t}) &\equiv
\cos \theta_{u_s,u_t} =
\frac{  \sigma_{u_s} \cdot \sigma_{u_t}  }{  ||\sigma_{u_s}||_2  ||\sigma_{u_t}||_2  } \\
sim(\iota_{v_s},\iota_{v_t}) &\equiv
\cos \theta_{v_s,v_t} =
\frac{  \iota_{v_s} \cdot \iota_{v_t}  }{  ||\iota_{v_s}||_2  ||\iota_{v_t}||_2  } 
\end{split}
\end{equation}

Therefore each element in the similarity matrix holds the edge weight for a pair of nodes, with a value ranging within the interval $[0,1]$. Edges weights are zero-valued if no direct link exist between two nodes,  whilst nodes linked by edges with a weight of $1$ have identical species bags or quorum vectors. Intermediate values reflect the correlation measure obtained for node pairs.
As this rule outputs weight values that are bound to a known interval, filtering less relevant links becomes much more straightforward. Depending on the aspects regarding the species-collectors system an investigator might be interested in, a filtering threshold $\phi$ can be set based on the minimum correlation value she considers acceptable, such that only the most relevant relationships for that particular analysis are kept.
%%% check neighborhood similarity functions in literature
%% Edges pruning


\subsection{Similar networked systems}
A variety of network models structurally similar to species-collectors networks have been proposed in social networks literature for modeling networked systems within many unrelated application domains. We describe some of them for improving the reader's intuition.

\textit{Affiliation networks} are a particular type of social networks in which actors are associated to events, groups or institutions by being members or participating in them \cite{Borgatti2015}. Social ties between individuals are usually derived from such models by considering that individuals belonging to the same groups or attending the same events are more likely to become acquainted than those who are not co-affiliated to any of those. 
A classic example of this kind of network is the movie actors network, which was used as an empirical example in \textit{Watts} and \textit{Strogratz}'s work \cite{Watts1998} while describing the small-world property in real-world networks.
This network was built from the Internet Movie Database (IMDB \footnote{\url{http://www.imdb.com/}}), linking actors to movies they have starred in.

% exemplify models

Although the semantics of relationships modeled in each of these networks (with different analytic purposes), the structures used to represent them are shared. A common feature shared by those models is their bipartite structure, allowing  

The bipartite structure of SCNs make them structurally similar to affiliation networks, with some examples in the social networks literature.

One classic affiliation network model is the films and actors network.

Lambiotte models listeners and music

Another analogous example are the scientific papers coauthorship networks in which authors are linked to papers they have authored \cite{Newman2004}



Whereas SCNs represent ..., other structures might arise when analyzing which collectors record together. 
The model presented in the next section covers this.










% =============================
% =============================
% Collectors Coworking Networks
% -----------------------------

\section{Collectors Coworking Networks}
% References: (Ramasco2004)
% TODO: Talk about singleton collectors
In this section we describe Collectors Coworking Networks (CWNs) as relational models for representing direct collaborative associations between collectors.

\subsection{General Description}
Collectors Coworking Networks are a particular instance of \textit{collaboration networks} \cite{Ramasco2004} describing coauthoring relationships between collectors from species occurrence records (Figure \ref{fig:cwn_general}).
We consider two collectors to be coauthors in a given record if they are both included in the collectors field for that record. The collectors field holds a list collectors names who have authored each record in the dataset, and is equivalent to the \textit{recordedBy} field in a dataset following \textit{Darwin Core} terms standards (check an example in Table \ref{table:cwn_example_dataset}). We refer to each distinct list of collectors in this context as a \textit{team}.

As opposed to SCNs, which describe collectors interests towards species, relationships in CWNs are directly formed between collectors who have effectively worked collaboratively in field, and are semantically described as ``\textbf{collector} records specimen with \textbf{collector}''.
Each individual species occurrence record with at least two collectors (team size greater than $1$) is thus considered a distinct collaboration act, originating new pairwise connections between all collectors involved.
For records with team size equal to $1$, which we refer to as non-collaborative records, no connections are created.

  \begin{figure}[h!]
  	\centering
    \includegraphics[width=.4\linewidth]{figures/network_models/cwn_generalaspect.pdf}
    \caption{General aspect of a Collectors Coworking Network (CWN). Coauthoring relationships between collectors (green nodes) are structured as edges in the graph, and are graphically represented as gray links.
    Here collectors nodes are sized according to the total number of records they've authored; and edges are weighted according to the number of collaborative records coauthored by both nodes.}
    \label{fig:cwn_general}
  \end{figure}

Differently from SCNs, where two entities classes are represented in the graph as disjoint nodes sets with the bipartite constraint, CWNs exclusively model direct relationships between entities from a single class (collectors), with the only connectivity restriction that a collector should not hold collaborative ties to itself.
The model is thus formally described as an unipartite (or one-mode) undirected graph
$$CWN = (S,E) \mbox{ ,}$$
where $S=\{u_1,u_2,...u_n\}$ is the graph's nodes set and $E$ is the set of undirected edges linking members of $S$.

Analogously to SCNs, both nodes and edges hold homonymous but conceptually distinct \textit{count} attributes.
Nodes count attribute stores the total number of records --- including non-collaborative ones --- a collector has authored, whereas
edges count attribute stores the total number of times an association between two collectors was observed in the dataset. Thus, although each node and edge respectively are uniquely included in $S$ and $E$, their recurrence patterns are registered in the model as attributes.
Another important edge attribute is the \textit{species list}. Although species associated to each occurrence record are not represented as entities in this model, edges can optionally keep a list of species that are shared by two collectors through that link, which is stored in this attribute.

Weights are assigned to edges in the CWN as a measure of their overall relevance in the network structure. Edges with higher weight values represent stronger collaborative ties between collectors, pointing out the main groups of collectors who are most willing to collaborate. 
The simplest rule is to set the edge weight as the total total number of occurrences of the tie it represents in the dataset. However, as pointed out by other authors studying social networks \cite{Newman2001a}, in reality not all collaboration acts should contribute the same way for a collector's network. 
Collectors tend to hold weaker collaborative ties with each other when they collaborate in larger teams than when they collaborate in smaller ones. 

The \textbf{hyperbolic weighting} rule accounts for this fact, while also considering the total number of collaborations between two collectors as a factor contributing the strength of their link.
According to this rule, not every new occurrence of the link increases the edge weight equally. 
The contribution of each new link depends on the number of the collectors $n^{(k)}$ included in record $k$ or, in other words, the team size. 
This rule follows a hyperbolic growth function 
\begin{equation}
w_{(i,j)} = \sum\limits_k \frac{\delta_i^{(k)} \delta_j^{(k)}}{(n^{(k)}-1)} \mbox{ , }
\end{equation}
where $\delta^{(k)}_u = 1$ if collector $u$ is in record $k$ and $0$ otherwise.
As the hyperbolic function above has singularity at $1$, it gets ill-defined for records with only one collector. 
Therefore only records with two or more collectors are used to compute edges weights. 
The maximum weight contribution of 1 is assigned to records with two collectors, whilst records with larger cliques yield smaller contributions.\\

Relationships in the CWN graph can be represented in a symmetric adjacency matrix $A^{n\times n}$ for which $a_{ij} \neq 0 \textit{ iff } (u_i,u_j) \in E$. 
Values of non-zero elements depend on the weighting method adopted for representing links strength, being the absolute counts of edges' recurrence (edges' \textit{count} attribute) the simplest one.
Additionally, the model's connectivity constraint states that all diagonal elements in $A$ are necessarily equal to $0$, thus ensuring that no self loops are formed.
To give a concrete example, the adjacency matrix for the graph in Figure \ref{fig:cwn_general} is 
$$
A =
\kbordermatrix{
& col1 & col2 & col3 & col4 & col5 & col6 & col7 & col8 & col9 & col10 \\
col1 & 0 & 2 & 2 & 0 & 1 & 0 & 0 & 0 & 0 & 0\\
col2 & 2 & 0 & 3 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\
col3 & 2 & 3 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\
col4 & 0 & 0 & 0 & 0 & 1 & 0 & 0 & 2 & 2 & 0\\
col5 & 1 & 0 & 0 & 1 & 0 & 1 & 1 & 0 & 0 & 0\\
col6 & 0 & 0 & 0 & 0 & 1 & 0 & 2 & 0 & 0 & 0\\
col7 & 0 & 0 & 0 & 0 & 1 & 2 & 0 & 0 & 0 & 0\\
col8 & 0 & 0 & 0 & 2 & 0 & 0 & 0 & 0 & 2 & 0\\
col9 & 0 & 0 & 0 & 2 & 0 & 0 & 0 & 2 & 0 & 0\\
col10 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0
},
$$
where each element is the count of the total number of recurrences of collectors associations, represented in the graph as edges weights.




% Network construction
%% Singleton collectors are those holding no collaborative recording with any other collector in the dataset ($k=0$), and are also included in the model as isolated nodes.
\subsection{Model construction from data}

We build CWNs from species occurrence data in an iterative process that is similar to the one described for SCNs.
In this case, however, the only field that is strictly required for structuring relationships is the one containing collectors names, which in a database following \textit{Darwin core} terms standards should be named ``\textit{recordedBy}''.
The field containing species identities, although not required, can be optionally used during model construction in case the user decides to set the edges' \textit{species list} attribute.
Table \ref{table:cwn_example_dataset} shows the species occurrence dataset which was used to build the graph in Figure \ref{fig:cwn_general}.

\begin{table}[!ht]
  \caption{Species occurrence dataset from which the CWN model in Figure \ref{fig:cwn_general} was built. The \textit{species} field is not strictly required for building CWN models.}
  \begin{center}
  \begin{tabular}{r l c}
      id & recordedBy & species \\
      \hline
        0 & col1; col2; col3 & sp1\\ 
        1 & col3; col1; col2 & sp2\\ 
        2 & col1; col3 & sp3\\ 
        3 & col5; col4 & sp3\\ 
        4 & col5; col2 & sp3\\ 
        5 & col5; col6 & sp5\\ 
        6 & col5; col7 & sp4\\ 
        7 & col6; col7 & sp6\\ 
        8 & col6; col7 & sp7\\ 
        9 & col4; col8; col9 & sp4\\ 
        10 & col4; col9; col8 & sp5\\ 
        11 & col10 & sp6\\ 
        12 & col10 & sp6\\
       \hline
  \end{tabular}
  \end{center}
  \label{table:cwn_example_dataset}
\end{table}

For each row in the dataset, a \textit{clique} structure is formed by creating edges between all collectors included in the record's team, in a pairwise fashion. 
Each clique thus represents one collaborative act, where every collector gets and additional collaborative tie with every other collector included in that collaborative record. 
The clique size, which is the number of nodes included in the clique structure, is equivalent to the team size. For non-collaborative records the clique is composed by the collector node itself, and therefore no edges are formed.
As the user might want to distinguish the relevance of links originated from distinct team sizes, the hyperbolic weighting rule described in the previous section can be used for weighting links in each clique. 
In case the species field is included in the building routine, each clique also gets associated to the name of the species to which the recorded specimen belongs to.

The CWN model is finally composed by combining all cliques together into a single undirected graph. In this process edges that occur in multiple cliques have their weights summed. In case the species list attribute is set, combining edges also merges their respective species lists.




% Are unconnected collector represented in the model?
%% Average number of collaborators?


%% Weighting
%%% Check the weighting rule from Newman2001a -> Hyperbolic weighting
